waffle-chai/src/matchers/ethCalled.ts

```diff
@@ -0,0 +1,56 @@
+import {Contract} from 'ethers';
+import {MockProvider, RecordedCall} from '@ethereum-waffle/provider';
+
+interface AssertionResultInterface {
+  wasCalled: boolean;
+  errorMessage: string;
+  negatedErrorMessage: string;
+}
+
+type IsInCallHistory = (call: RecordedCall) => boolean
+
+export function supportEthCalled(Assertion: Chai.AssertionStatic) {
+  Assertion.addProperty('ethCalled', function () {
+    const arg = this._obj;
+    const [contract, fnName] = Array.isArray(arg) ? arg : [arg];
+
+    if (!(contract instanceof Contract)) {
+      throw new TypeError('ethCalled: argument must be a contract');
+    }
+    const provider = contract.provider;
+    if (!(provider instanceof MockProvider)) {
+      throw new TypeError('ethCalled: contract.provider must be a MockProvider');
+    }
+    if (fnName !== undefined && typeof fnName !== 'string') {
+      throw new TypeError('ethCalled: function name must be a string');
+    }
+
+    const fnSighash = fnName && contract.interface.functions[fnName].sighash;
+
+    const isInCallHistory = (cb: IsInCallHistory): boolean => provider.callHistory.some(cb);
+
+    const getAssertionResult = (): AssertionResultInterface => {
```

Comment:
No need to have a function here. You could do an if on `fnSighash` and do different asserts in `if` and `else`. The code will be cleaner and shorter